src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs

```diff
@@ -248,6 +225,49 @@ void ExecuteWithAbi (string [] supportedAbis, string apkInputPath, string apkOut
 						count = 0;
 					}
 				}
+
+				HashSet<ulong> deletedEntries = new HashSet<ulong> ();
+				if (apkInputPathExists) {
+					using (var packaged = new ZipArchiveEx (apkInputPath, FileMode.Open)) {
+						foreach (var entry in packaged.Archive) {
+							Log.LogDebugMessage ($"Deregistering item {entry.FullName}");
+							existingEntries.Remove (entry.FullName);
+							if (lastWriteInput <= lastWriteOutput)
+								continue;
+
+							long entryIndexInOutput = -1;
+							if (apk.Archive.ContainsEntry (entry.FullName, out entryIndexInOutput)) {
+								ZipEntry e = apk.Archive.ReadEntry (entry.FullName);
+								// check the CRC values as the ModifiedDate is always 01/01/1980 in the aapt generated file.
+								if (entry.CRC == e.CRC) {
+									Log.LogDebugMessage ($"Skipping {entry.FullName} from {apkInputPath} as its up to date.");
+									continue;
+								}
+							}
+
+							var ms = new MemoryStream ();
+							entry.Extract (ms);
+
+							if (entryIndexInOutput != -1) {
+								Log.LogDebugMessage ($"Refreshing {entry.FullName} from {apkInputPath}");
+
+								// Force the modified resource to move to the end of the file, by deleting it first so that AddStream adds it
+								// back with a new index. Keeping modified resources toward the end optimizes the delta install for the typical
+								// dev scenario where the user is editing a few resources but most of the APK contents (e.g. the native libs)
+								// don't change and we want to keep their byte offset in the APK fixed. Delta install need not update APK
+								// contents that don't change and don't move.
+								apk.Archive.DeleteEntry ((ulong) entryIndexInOutput);
```

Comment:
@grendello: I don't understand the semantics here.  What is `entryIndexInOutput`. It comes from `(apk.Archive.ContainsEntry(entry.FullName, out entryIndexInOutput)`, but *what is it*, really?

An "index in the zip file", presumably?

Furthermore, why doesn't `apk.Archive.DeleteEntry()` *invalidate* any previously obtained `entryIndexInOutput` values?  Is `entryIndexInOutput` guaranteed to always increase, even when "previous" entries in the zip have been removed?

This implies that previously obtained `entryIndexInOutput` values won't change, otherwise it wouldn't work: https://github.com/xamarin/xamarin-android/pull/4659/files#diff-6502ef3162df5eefc5426881ce336007R131

But if that's the case, when *do* the `entryIndexInOutput` values change?  When the file is *saved*?

Doesn't `ZipArchiveEx` *implicitly* close and re-open the underlying zip "occasionally"?  See e.g. 296d8544d04e3ee9ad513689eb48db30d2f716d9.

https://github.com/xamarin/xamarin-android/blob/3f438e46d7b166a3a3ef54c9ffafb5f426760468/src/Xamarin.Android.Build.Tasks/Utilities/ZipArchiveEx.cs#L30-L38
https://github.com/xamarin/xamarin-android/blob/3f438e46d7b166a3a3ef54c9ffafb5f426760468/src/Xamarin.Android.Build.Tasks/Utilities/ZipArchiveEx.cs#L72-L75
I *really* don't fully understand the intended semantics here, but what I think I understand, I'm not sure I'm comfortable with.

Reply:
Answering some of those questions:

Yes the index matches the "index in the zip" and always increases (until the zip is written out). The index is managed by the underlying C libzip code. This is basically the algorithm, from what I observed: When the zip is opened, the index matches the position in the zip file (0 is at the beginning, the max value at the end). While the zip is being modified in memory by libzip, new entries are assigned an index of "previousMaxIndex + 1", so the index always increases and deleted entries keep their current index number (so there's a hole). And then when the zip is written & closed, entries are written in index order with deleted entries skipped, so when the file is reopened the index numbers can be different than before (since deleted entries have their index number reused by whatever non-deleted entry came after - no more holes).

FYI, I was using `unzip -Zv <zipfile>` to dump the index of the zip and see what order things are in. The index number reported by `unzip -Zv` is the same as managed by libzip, except that unzip reports indexes as being 1 based while libzip has them as 0 based. But conceptually they are the same, indicating the relative order of entries in the zip.

And yes, Flush can change the index numbers, but fortunately it's not called here.

Anyway, all that said, it would probably be cleaner to update LibZipSharp so that the enumerator can cope with deletions, skipping them instead of throwing an exception. Adding a ReadEntry overload that doesn't throw on deleted items and then calling it from https://github.com/xamarin/LibZipSharp/blob/a0973d4a861bf683cf225bc67111b249999673ce/ZipEntryEnumerator.cs#L59 is probably the best way to do that.

@grendello - would you want to take a look at making that change?

Reply:
@jonpryor what @BretJohnson described is what happens. Basically, `libzip` maintains only an in-memory image of the zip archive which, until flushed, won't cause problems with indexes being changed underneath. 

@BretJohnson I'd be happy to review a PR, frankly, at this moment in time... :) So unless it can wait, I'm not the right person to do it right now... :)

Reply:
@BretJohnson wrote:

> And yes, Flush can change the index numbers, but fortunately it's not called here.

*`ZipArchiveEx.AddFiles()` calls `Flush()`*.  *Adding* a file *may* call `Flush()`.

Which means for the entire "scope" between `deletedEntries = new HashSet…` and `apk.FixupWindowsPathSeparators(deletedEntries…)`, `apk.Archive.Add*()` *must not be called*.

It isn't *currently* called, but this constraint is not at all obvious, and thus could be easily broken in the future.

*At minimum* that "scope" should be moved to a new method, with adequate documentation that no entries can be added to `apk.Archive`.

"Better" would be to remove the need for `deletedEntries` in the first place, which would require libZipSharp changes, but that's not necessarily a bad thing.

Reply:
@grendello @jonpryor - I updated LibZipSharp to handle this. Can you please review that PR, here: https://github.com/xamarin/LibZipSharp/pull/53

After we merge that we'll bump libZipSharp in this PR.

Reply:
I removed this workaround code, with the LibZipSharp update to now skip deleted entries when enumerating. And I bumped xamarin-android to use the new LibZipSharp version 1.0.12, which isn't (yet) published.

@dellis1972 - Can you publish the new LibZipSharp release, from the latest LibZipSharp commit. Again, it should be 1.0.12 (as I bumped the version on the LibZipSharp side). That should be the last thing needed here.

Reply:
1.0.13 of LibZipSharp has been published